### PR TITLE
Further improve `impl Trait`/`dyn Trait` suggestions

### PR DESCRIPTION
**Pull request by estebank**
After reading [_Returning Trait Objects_ by Bryce Fisher-Fleig](https://bryce.fisher-fleig.org/blog/returning-trait-objects/), [I noticed that](https://www.reddit.com/r/rust/comments/esueur/returning_trait_objects/ffczl4k/) #68195 had a few bugs due to not ignoring `ty::Error`.

- Account for `ty::Error`.
- Account for `if`/`else` and `match` blocks when pointing at return types and referencing their types.
- Increase the multiline suggestion output from 6 lines to 20.

### CODE REVIEW

**Comment by estebank** [comment]
CC @oli-obk

File: src/librustc/traits/error_reporting/suggestions.rs [hunk 1]
```diff
@@ -1699,20 +1715,41 @@ impl<'v> Visitor<'v> for ReturnsVisitor<'v> {
     }
 
     fn visit_expr(&mut self, ex: &'v hir::Expr<'v>) {
-        if let hir::ExprKind::Ret(Some(ex)) = ex.kind {
-            self.0.push(ex);
+        match ex.kind {
```
Comment (Centril) [inline src/librustc/traits/error_reporting/suggestions.rs]: The changes here feel unclear to me re. what is being achieves. Could you leave more comments?

File: src/librustc/traits/error_reporting/suggestions.rs [hunk 2]
```diff
@@ -1701,20 +1711,41 @@ impl<'v> Visitor<'v> for ReturnsVisitor<'v> {
...
+            }
+            // We need to walk to find `return`s in the entire body.
+            _ if !self.in_block_tail => hir::intravisit::walk_expr(self, ex),
+            _ => self.returns.push(ex),
         }
-        hir::intravisit::walk_expr(self, ex);
     }
 
     fn visit_body(&mut self, body: &'v hir::Body<'v>) {
+        let prev = self.in_block_tail;
```
Comment (oli-obk) [inline src/librustc/traits/error_reporting/suggestions.rs]: shouldn't this always be false? Entering an inner body should not be possible (due to `hir::intravisit::NestedVisitorMap::None`, so you can `assert!(!self.in_block_tail);` here

File: src/librustc/traits/error_reporting/suggestions.rs [hunk 3]
```diff
@@ -1701,20 +1711,41 @@ impl<'v> Visitor<'v> for ReturnsVisitor<'v> {
...
             if let hir::ExprKind::Block(block, None) = body.value.kind {
-                if let Some(expr) = block.expr {
-                    self.0.push(expr);
+                if block.expr.is_some() {
+                    self.in_block_tail = true;
                 }
             }
         }
         hir::intravisit::walk_body(self, body);
+        self.in_block_tail = prev;
```
Comment (oli-obk) [inline src/librustc/traits/error_reporting/suggestions.rs]: this is unnecessary out of the same reason given for the creation of `prev`

File: src/test/ui/error-codes/E0746.rs [hunk 4]
```diff
@@ -10,9 +10,9 @@ fn foo() -> dyn Trait { Struct }
 
 fn bar() -> dyn Trait { //~ ERROR E0746
     if true {
-        return 0;
+        return 0u32;
```
Comment (oli-obk) [inline src/test/ui/error-codes/E0746.rs]: why are these changes necessary? What happens if you don't do it?

File: src/librustc_errors/emitter.rs [hunk 5]
```diff
@@ -456,9 +456,14 @@ impl Emitter for SilentEmitter {
...
 }
 
-/// maximum number of lines we will print for each error; arbitrary.
+/// Maximum number of lines we will print for each error; arbitrary.
 pub const MAX_HIGHLIGHT_LINES: usize = 6;
-/// maximum number of suggestions to be shown
+/// Maximum number of lines we will print for a multiline suggestion; arbitrary.
+///
+/// This should be replaced with a more involved mechanism to output multiline suggestions that
+/// more closely mimmics the regular diagnostic output, where irrelevant code lines are ellided.
```
Comment (oli-obk) [inline src/librustc_errors/emitter.rs]:
```suggestion
/// more closely mimmics the regular diagnostic output, where irrelevant code lines are elided.
```

File: src/test/ui/error-codes/E0746.rs (hunk 4 above)
Comment (estebank) [reply src/test/ui/error-codes/E0746.rs]: This was needed before I did `resolve_vars_if_possible` because `{integer}` wasnt being detected as implementing `Trait`. Now it does.

Comment (oli-obk) [reply src/test/ui/error-codes/E0746.rs]: So these changes can be reverted?

File: src/librustc/traits/error_reporting/suggestions.rs (hunk 1 above)
Comment (estebank) [reply src/librustc/traits/error_reporting/suggestions.rs]: All of this is to be able to point at the leafs of the return path graph, not just at the tail expr (which can be match/if-else).

File: src/test/ui/error-codes/E0746.rs (hunk 4 above)
Comment (estebank) [reply src/test/ui/error-codes/E0746.rs]: yes

File: src/librustc_errors/emitter.rs [hunk 6]
```diff
@@ -1521,7 +1526,7 @@ impl EmitterWriter {
             draw_col_separator_no_space(&mut buffer, 1, max_line_num_len + 1);
             let mut line_pos = 0;
             let mut lines = complete.lines();
-            for line in lines.by_ref().take(MAX_HIGHLIGHT_LINES) {
+            for line in lines.by_ref().take(MAX_SUGGESTION_HIGHLIGHT_LINES) {
```
Comment (oli-obk) [inline src/librustc_errors/emitter.rs]: This is highly likely to break clippy's ui tests, let's hold up until after the breakage week

**Comment by estebank** [comment]
@bors r=oli-obk
reverted the suggestion output change
